### PR TITLE
magic word

### PR DESCRIPTION
**Pull request by zvezdochiot**
* added magic word (dirty)
* ~~astyle --style=allman~~

### CODE REVIEW

**Comment by zvezdochiot** [comment]
I propose a forcing dirty `magic` processing scheme, because it is completely unclear to me how to process the file when the` magic` is damaged.

**Review by fridex — COMMENTED** [review]
> I propose a forcing dirty `magic` processing scheme, because it is completely unclear to me how to process the file when the` magic` is damaged.

What about creating a new parameter that would force the magic from the command line? Otherwise it can read from the magic.

BTW could you please open a separate PR for code style changes?

**Comment by zvezdochiot** [comment]
@fridex say:
>What about creating a new parameter that would force the magic from the command line? Otherwise it can read from the magic.

No need. In case of damage `magick` `blocksize = 0 -> RET_ERR_ARGS`. To work around this, you must specify the length of the block directly:
```sh
./rsdec LICENSE.rsecc 1024
```

@fridex say:
>BTW could you please open a separate PR for code style changes?

How to do it now?

**Comment by fridex** [comment]
> @fridex say:
> 
> > What about creating a new parameter that would force the magic from the command line? Otherwise it can read from the magic.
> 
> No need. In case of damage `magick` `blocksize = 0 -> RET_ERR_ARGS`. To work around this, you must specify the length of the block directly:
> 
> ```shell
> ./rsdec LICENSE.rsecc 1024
> ```

Indeed. Besides that, I would expect also a message to the user that the magic was not parsed and should be supplied explicitly.

> @fridex say:
> 
> > BTW could you please open a separate PR for code style changes?
> 
> How to do it now?

Create a new branch with the coding style changes, [1] might be useful.

[1] https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request

**Comment by zvezdochiot** [comment]
@fridex say:
>Indeed. Besides that, I would expect also a message to the user that the magic was not parsed and should be supplied explicitly.

Not sure if this will be a complete construction.
If magic is damaged, is a message enough? What should be the message so that the user does not panic? How should a decoder behave if a parameter is missing?

Immediately rejected the idea of how to resolve these issues given PR. All these issues require a much more thorough study.

@fridex say:
>Create a new branch

Until I figured it out.

**Comment by zvezdochiot** [comment]
@fridex . Given the variety of possible solutions, I see the point in using my practices for YOUR solution to the question. That would be more correct, since I have more questions than answers.

**Comment by fridex** [comment]
> @fridex say:
> 
> > Indeed. Besides that, I would expect also a message to the user that the magic was not parsed and should be supplied explicitly.
> 
> Not sure if this will be a complete construction.
> If magic is damaged, is a message enough? What should be the message so that the user does not panic? How should a decoder behave if a parameter is missing?

From user's point of view, I would expect the following behavior:

1. if magic looks ok, use it implicitly if the parameter is not provided (implicit magic use assuming it was not damaged in a bad way)
2. if magic is damaged, print it out and fail with an error message if the parameter is not provided (cannot determine the parameter)
3. if the parameter is provided explicitly, use it - if it does not match the magic, log a warning

This way a user can use the program sanely - force the parameter if she/he wants to or use the implicit one parsed from the magic if required/requested so.

**Comment by zvezdochiot** [comment]
>From user's point of view, I would expect the following behavior:

>if magic looks ok, use it implicitly if the parameter is not provided (implicit magic use assuming it was not damaged in a bad way)

https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L248-L257

>if magic is damaged, print it out and fail with an error message if the parameter is not provided (cannot determine the parameter)

https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L250-L254
https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L258-L264

>if the parameter is provided explicitly, use it - if it does not match the magic, log a warning

Hmm. Warning text? (my English is bad).

File: src/coder.cpp [hunk 1]
```diff
@@ -237,6 +236,23 @@ int main(int argc, char * argv[]) {
...
+    if (! file_in.is_open()) {
+        std::cerr << fname.c_str() << ":"
+                  << strerror(errno) << std::endl;
+        return RET_ERR_FILE_IN;
+    }
+
+#ifdef BMS1B
+    char mc[10] = {"\0"};
+    file_in.read(mc, 9);
+    if (!sscanf(mc, "RSECC%d", &blocksize))
```
Comment (fridex) [inline src/coder.cpp]: This will not work properly in case of EOF that can be returned. maybe:

```C++
    if (sscanf(mc, "RSECC%d", &blocksize) <= 0)
```

**Comment by fridex** [comment]
> > From user's point of view, I would expect the following behavior:
> 
> > if magic looks ok, use it implicitly if the parameter is not provided (implicit magic use assuming it was not damaged in a bad way)
> 
> https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L248-L257
> 
> > if magic is damaged, print it out and fail with an error message if the parameter is not provided (cannot determine the parameter)
> 
> https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L250-L254
> https://github.com/FS-make-simple/rs-coder/blob/14d70c6b1659446b478ab59fffa33317e5b98d9a/src/coder.cpp#L258-L264
> 
> > if the parameter is provided explicitly, use it - if it does not match the magic, log a warning
> 
> Hmm. Warning text? (my English is bad).

No worries, mine too. What about:

```The block size parsed from the file magic %d does not correspond to the one supplied explicitly %d, using the explicit block size %d```

**Comment by zvezdochiot** [comment]
Resolved.

**Review by fridex — APPROVED** [review]
Awesome! Thanks! :+1:

**Comment by zvezdochiot** [comment]
It's too early. Test, test, test, ...

**Pull request merged**